File: tools/deltalog-bench/deltalog_bench.c
/* deltalog_bench.c -- consumer-side micro-benchmark driver for the
 * KaRaMeL-extracted C build of RDF.Store.Columnar.DeltaLog
 * (formal/fstar/c-output/deltalog/Factoidal_DeltaLog.{c,h}).
 *
 * Per CLAUDE.md rule #11 this is a consumer tool, not part of the
 * verified library: it contains NO byte-layout logic of its own,
 * only calls into the F*-extracted serialize_delta_batch /
 * parse_delta_batch functions the same way
 * formal/fstar/c-output/deltalog/demo/delta_log_demo.c does (that
 * file is the correctness proof; this one is its timing sibling).
 *
 * Builds a delta_batch of N synthetic DE_Add entries
 * (http://example.org/sI knows http://example.org/oI, no graph),
 * then times:
 *   - serialize_delta_batch(batch)            -> bytes
 *   - parse_delta_batch(bytes)                -> Some (batch', rest)
 * with clock_gettime(CLOCK_MONOTONIC). Prints one JSON line to stdout.
 *
 * Compiles unmodified for both native (gcc/clang) and wasm32-wasi
 * (clang --target=wasm32-wasi --sysroot=/usr, using the apt packages
 * wasi-libc + libclang-rt-18-dev-wasm32) -- see
 * tools/bench-runtimes.sh for both build recipes and
 * docs/web/perf/index.md for what that answers about a hypothetical
 * KaRaMeL C -> wasm full-engine build.
 *
 * Usage: deltalog_bench N
 *   prints: {"n":N,"bytes":B,"serialize_s":F,"parse_s":F,"parse_ok":true|false}
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "Factoidal_DeltaLog.h"
#include "krmlinit.h"

typedef struct Prims_list__RDF_Store_Columnar_DeltaLog_delta_entry_s entry_list;
typedef struct Prims_list__FStar_Char_char_s byte_list;

static double now_seconds(void) {
  struct timespec ts;
  clock_gettime(CLOCK_MONOTONIC, &ts);
  return (double)ts.tv_sec + (double)ts.tv_nsec / 1e9;
}

static RDF_Store_Columnar_DeltaLog_delta_entry mk_entry(size_t i) {
  char *s = (char *)malloc(48);
  char *o = (char *)malloc(48);
  snprintf(s, 48, "http://example.org/s%zu", i);
  snprintf(o, 48, "http://example.org/o%zu", i);

  RDF_Term_subject subj;
  subj.tag = RDF_Term_S_IRI;
  subj.case_S_IRI = (Prims_string)s;

  RDF_Term_rdf_term obj;
  obj.tag = RDF_Term_T_IRI;
  obj.case_T_IRI = (Prims_string)o;

  RDF_Triple_triple tr;
  tr.s = subj;
  tr.p = (Prims_string)"http://xmlns.com/foaf/0.1/knows";
  tr.o = obj;

  FStar_Pervasives_Native_option__Prims_string g;
  g.tag = FStar_Pervasives_Native_None;

  RDF_Store_Columnar_DeltaLog_delta_entry e;
  e.tag = RDF_Store_Columnar_DeltaLog_DE_Add;
  e.case_DE_Add.quad = tr;
  e.case_DE_Add.graph = g;
  return e;
}

int main(int argc, char **argv) {
  krmlinit_globals();

  if (argc < 2) {
    fprintf(stderr, "usage: %s N\n", argv[0]);
    return 2;
  }
  size_t n = (size_t)strtoull(argv[1], NULL, 10);

  /* Build the list back-to-front so no `append`/reverse of a
   * growing list is needed here (this driver's own cost, not the
   * measured operation's). */
  entry_list *acc = (entry_list *)malloc(sizeof(entry_list));
  acc->tag = Prims_Nil;
  for (size_t i = n; i > 0; i--) {
    entry_list *node = (entry_list *)malloc(sizeof(entry_list));
    node->tag = Prims_Cons;
    node->hd = mk_entry(i - 1);
    node->tl = acc;
    acc = node;
  }

  RDF_Store_Columnar_DeltaLog_delta_batch batch;
  batch.db_seq = 1;
  batch.db_epoch = 0;
  batch.db_ops = acc;

  double t0 = now_seconds();
  byte_list *bytes =
      (byte_list *)RDF_Store_Columnar_DeltaLog_serialize_delta_batch(batch);
  double t1 = now_seconds();

  size_t nbytes = 0;
  for (byte_list *p = bytes; p->tag == Prims_Cons; p = p->tl) nbytes++;

  double t2 = now_seconds();
  FStar_Pervasives_Native_option___RDF_Store_Columnar_DeltaLog_delta_batch___Prims_list__FStar_Char_char_
      parsed = RDF_Store_Columnar_DeltaLog_parse_delta_batch(
          (Prims_list__FStar_Char_char *)bytes);
  double t3 = now_seconds();

  int ok = parsed.tag == FStar_Pervasives_Native_Some;

  printf(
      "{\"n\":%zu,\"bytes\":%zu,\"serialize_s\":%.6f,\"parse_s\":%.6f,"
      "\"parse_ok\":%s}\n",
      n, nbytes, t1 - t0, t3 - t2, ok ? "true" : "false");

  return ok ? 0 : 1;
}

File: tools/deltalog-bench/deltalog_bench.ml
(* deltalog_bench.ml -- consumer-side micro-benchmark driver, native
   OCaml sibling of tools/deltalog-bench/deltalog_bench.c (the KaRaMeL
   C build) and the js_of_ocaml/wasm_of_ocaml `deltaBatchToHex` path
   (bin/npm-entry/entry_jsoo.ml).

   Per CLAUDE.md rule #11: consumer tool, no byte-layout logic of its
   own. Two modes calling only F*-extracted functions:

     pure N
       Build N DE_Add delta_entry ops directly (same shape as
       bin/delta-log-probe/probe.ml's batch_for_seq / the C harness's
       mk_entry), time RDF_Store_Columnar_DeltaLog.serialize_delta_batch
       and .parse_delta_batch. Comparable to the C-native/C-wasm
       numbers (no SPARQL parsing involved).

     sparql N
       Build one `INSERT DATA { ... N triples ... }` SPARQL Update
       string, time the exact same pipeline
       bin/npm-entry/entry_jsoo.ml's `delta_batch_to_hex` runs:
       SPARQL11_Parser.parse_sparql_update ->
       RDF_Store_Columnar_DeltaMerge.update_ops_to_delta_entries ->
       RDF_Store_Columnar_DeltaLog.serialize_delta_batch. Comparable
       to the js_of_ocaml/wasm_of_ocaml `deltaBatchToHex` numbers.

   Usage: deltalog_bench (pure|sparql) N
   Prints one JSON line to stdout.
*)

module DL = RDF_Store_Columnar_DeltaLog
module DMerge = RDF_Store_Columnar_DeltaMerge

let now () = Unix.gettimeofday ()

(* ---- pure mode: hand-built delta_entry list, no SPARQL ---- *)

let build_ops_pure (n : int) : DL.delta_entry list =
  List.init n (fun i ->
    let s : RDF_Term.subject =
      RDF_Term.S_IRI (Printf.sprintf "http://example.org/s%d" i) in
    let p : Prims.string = "http://xmlns.com/foaf/0.1/knows" in
    let o : RDF_Term.rdf_term =
      RDF_Term.T_IRI (Printf.sprintf "http://example.org/o%d" i) in
    let tr : RDF_Triple.triple = { RDF_Triple.s; RDF_Triple.p; RDF_Triple.o } in
    DL.DE_Add (tr, None))

let run_pure (n : int) : unit =
  let ops = build_ops_pure n in
  let batch : DL.delta_batch =
    { DL.db_seq = Z.of_int 1; DL.db_epoch = Z.of_int 0; DL.db_ops = ops } in
  let t0 = now () in
  let bytes = DL.serialize_delta_batch batch in
  let t1 = now () in
  let nbytes = List.length bytes in
  let t2 = now () in
  let parsed = DL.parse_delta_batch bytes in
  let t3 = now () in
  let ok = match parsed with FStar_Pervasives_Native.Some _ -> true | _ -> false in
  Printf.printf
    "{\"mode\":\"pure\",\"n\":%d,\"bytes\":%d,\"serialize_s\":%.6f,\"parse_s\":%.6f,\"parse_ok\":%b}\n%!"
    n nbytes (t1 -. t0) (t3 -. t2) ok

(* ---- sparql mode: INSERT DATA text -> delta_batch_to_hex pipeline ---- *)

let build_insert_data_text (n : int) : string =
  let buf = Buffer.create (n * 64) in
  Buffer.add_string buf "INSERT DATA { ";
  for i = 0 to n - 1 do
    Buffer.add_string buf
      (Printf.sprintf "<http://example.org/s%d> <http://xmlns.com/foaf/0.1/knows> <http://example.org/o%d> . "
         i i)
  done;
  Buffer.add_string buf "}";
  Buffer.contents buf

let hex_of_bytes (bs : RDF_Bytes.bytes) : string =
  let buf = Buffer.create (List.length bs * 2) in
  List.iter (fun b -> Buffer.add_string buf (Printf.sprintf "%02x" b)) bs;
  Buffer.contents buf

let run_sparql (n : int) : unit =
  let text = build_insert_data_text n in
  let t0 = now () in
  let hex, op_count =
    match SPARQL11_Parser.parse_sparql_update text with
    | SPARQL11_Parser.ParseErr msg -> failwith ("SPARQL update parse error: " ^ msg)
    | SPARQL11_Parser.ParseOk (u, _rest) ->
      (match DMerge.update_ops_to_delta_entries "bench_0" u.SPARQL11_Algebra.u_ops with
       | FStar_Pervasives_Native.None -> failwith "unsupported update op"
       | FStar_Pervasives_Native.Some entries ->
         let batch : DL.delta_batch =
           { DL.db_seq = Z.of_int 1; DL.db_epoch = Z.of_int 0; DL.db_ops = entries } in
         let bytes = DL.serialize_delta_batch batch in
         (hex_of_bytes bytes, List.length entries))
  in
  let t1 = now () in
  Printf.printf
    "{\"mode\":\"sparql\",\"n\":%d,\"opCount\":%d,\"hexLen\":%d,\"total_s\":%.6f}\n%!"
    n op_count (String.length hex) (t1 -. t0)

let () =
  match Sys.argv with
  | [| _; "pure"; n |] -> run_pure (int_of_string n)
  | [| _; "sparql"; n |] -> run_sparql (int_of_string n)
  | _ ->
    prerr_endline "usage: deltalog_bench (pure|sparql) N";
    exit 2
